### ISSUE TITLE
How to verify Detached JWS signature 

### ISSUE THREAD

**Issue by tdtrivedi** [dvsekhvalnov/jose-jwt#151]
Hi, 

I am trying to verify a detached JWS signature and am not successful as yet. Any pointers will be helpful. 

I have a requirement where in I need to use symmetrically encrypted payload to create a token, make it a detached token and send it across to third-party. I will receive their detached tokens the same way. I need to verify these tokens based on the encrypted payload (which shall be body of the request). 

From what I understand, I can pass JwtOption for DetachPayload and I can get appropriate detached token. However, this does not include the payload (for obvious reasons based on the options) - which is what I do not want. In this case, I would need to get a full token and then detach it myself before sending it across with the encrypted body. 

All works well for creation - I can debug it successfully in JWT.io debugger. 

However, I am at a loss on how to verify it from my code. I am not able to successfully decode if I construct the JWT by following method. 

JWSDetached.split('.')[0] + '.' + Base64URL.Encode(PayloadBytes) + '.' + JWSDetached.split('.')[2] 

Any pointers on what I should be doing here ?

Thanks.

**Comment by dvsekhvalnov** [comment]
Hi @tdtrivedi , 

not sure i fully understood your use case. Documentation for RFC7797 supported options is here (may be you've already seen it): https://github.com/dvsekhvalnov/jose-jwt#unencoded-and-detached-content-aka-rfc-7797

if you can illustrate your question with some pseudo-code example - will be really great for understanding.

**Comment by tdtrivedi** [comment]
Hi @dvsekhvalnov ,

Thanks for the revert. 

Here is what I am trying to do.

Create a detached token which has encrypted content. The content will mostly be JSON (as of the use cases now). Part of the requirement for using detached token is to be able to carry out verification at the receiving end for any tampering with the payload - hence the need to rebuild the token fully at the receiving end to verify the signature. 

Token creation
`            Jose.JwtOptions options = new Jose.JwtOptions();`
`            options.DetachPayload = false;`
`            options.EncodePayload = false;`
`            JWTtoken = Jose.JWT.Encode( inputObj, privateKey, JwsAlgorithm.RS256, options: options );`

I was able to create the complete token and verify it using JWT.io debugger. When I use the DetachedPayload option, I get the header and the signature but as I see from the Encode() code in Jose.JWT - the payload is not taken as part of the signature creation. That defeats the purpose for which we want to use the Detached token. 

This is the reason why we are obtaining a regular JWT token and then building the detached token out of it. On the receiving side, we need to be able to rebuild the complete JWT token. Current thought about this is to obtain the detached token using x-jws-signature as part of the API call and the payload from the HTTP POST body. 

`            string completeToken = request.token.Split('.')[0] + "." + request.body.encryptedContent + "." + request.token.Split('.')[2];`
`            string verifiedPayload = Jose.JWT.Decode(completeToken, pubKey, JwsAlgorithm.RS256);`

When this is done, I get IntegrityException while using JWT.Decode() . 

I have seen the RFC 7797 implementation but as explained above, it does not help me. 

I was wondering where I am going wrong here ? Can you suggest something ?

Thanks

**Comment by dvsekhvalnov** [comment]
@tdtrivedi still not follow, when you set `DetachPayload=false` - you'll get normal 3 parts signed token, it's not detaching anything (and this is default option).

If you want to encode & detach you need to pass `DetachPayload=true` and library will produce token like:
```
eyJiNjQiOmZhbHNlLCJjcml0IjpbImI2NCJdLCJhbGciOiJSUzI1NiJ9..iyormYw6b0zKjx4K-fpeZO8xrLghkeUFMb2l4alz03CRLVdlXkdeKVG7N5lBbS-kXB4-8hH1ELFA5fUJzN2QYR6ZZIWjDF77HYTw7lsyjTJDNABjBFn-BIXlWatjNdgtRi2BZg2q_Wos87ZQT6Sl-h5hvxsFEsR0kGPMQ4Fjp-sxOyfnls8jAlziqmkpN-K6I3tK2vCLCQgnaN9sYrsIcrzuEA30YeXsgUe3m44yxLCXczXWKE3kgGiZ0MRpVvKOZt4B2DZLcRmNArhxjhWWd1nKZvv8c7kN0TqOjcNEUGWzwDs4ikCSz1aYKaLPXgjzpKnzbajUM117F3aCAaWH9g
```

, where 'middle' or 'content' part is stripped out.


If you want to verify such token with detached content, you have to explicitly provide missing payload to library:
```c#
Jose.JWT.Decode(token, someKey, payload: @"{""whatever payload"": ""was detached""}");
```
Does it make sense?

**Comment by tdtrivedi** [comment]
Hi @dvsekhvalnov 

Thanks for clarifying this for me. I have a comment and 2 followup questions.

Firstly, I should not have pasted the DetachPayload=False statement in my earlier post - that was part of my various debug attempts. It was misleading for my post - sorry about that. 

Secondly, I understand that if I pass detachpayload as false - i will get 3 part token. 

I tried your decode example with EncodePayload = False and was able to verify the signature after reconstructing the token based on x-jws-signature value and request body. 

I have two followup.
1. Just for my understanding - If I were to pass DetachPaylod = True - is the signature created with header only ? (this is my understanding currently)
2. When I use EncodePayload = False - the payload itself is enclosed in double quotes - this means when I verify - I need to do this specially - i.e. enclose my request.payload in double quotes before verification. Is there a way out of this ? The question here is if I am getting the payload from some third-party who is using some other JWT library, I may not be able to verify that content with the same method of enclosing in double quotes. 

Thanks

**Comment by dvsekhvalnov** [comment]
@tdtrivedi :

1. no, signature is always covers **payload + headers**. The only difference when using `detached` option - payload is not included resulting token.

2. didn't really get what do you mean by double quotes. library is never modifying payload at all, it's immutable. What `EncodePayload` options means is whether to apply **base64url** encoding for payload (default `true`). If you planning to include payload in a token you normally always want it or otherwise you can get invalid tokens (for instance unencoded payload can contain '.' characters, making decoding impossible).

The only real use case to omit encoding is when you want to use detached mode and your payload is big enough so you can save some CPU cycles by avoiding extra calculations.

**Comment by tdtrivedi** [comment]
Hi @dvsekhvalnov  ,

The double quotes issue - it was due to the fact that my helper class which actually generates the token was taking the payload as Object which somehow resulted in the payload enclosed in double quotes. Once I changed it to a String, things are working fine now. I have not had the time to decipher why its behaving like that !! 

E.g.  
If I pass the payload as Object - this is what is being generated as JWT. Notice the payload enclosed in double quotes. 

eyJhbGciOiJSUzI1NiIsImI2NCI6ZmFsc2UsImNyaXQiOlsiYjY0Il0sInR5cCI6IkpXVCJ9."LWv/pxykqNB7xrlRP7CIO7suNB96qkKfDo8Tff5
...redacted....
/31xvoynccITjkW4mg==".ayEhMz0y0vNWB8_La4uuWt8elQhVcDjmOFbZvdPkpjIYEvhMH1BHEE1us1b4XLD4i5qdnp-CssyJ50SNg-WYHHxqF48QhmW2FyBTOCRRjs4t7ZD-CNyHdha85_1QAp3or-BgBTKUaUn4fFamMJW-6V4_huxWMHP8iK__QueLI_HeLcZHdfXkHsottM4iBjENT2gKKzCBZUKL59BZmPeldtBqN098E5pleXFI1LhOFEQldlqFakl-wrqw-8q-dFisBf5Ht8CAbWiQ7BQT5I6n4FQe7afwR_vuj-u4WLsCYZ_F71m9OxV83er2fQ2W4S4O2nj4F6_9CIOyXCWgf_6XSg


But if I send out the payload as string .. the same token is generated without the quotes. 

eyJhbGciOiJSUzI1NiIsImI2NCI6ZmFsc2UsImNyaXQiOlsiYjY0Il0sInR5cCI6IkpXVCJ9.LWv/pxykqNB7xrlRP7CIO7suNB96qkKfDo8Tff5
...redacted....
/31xvoynccITjkW4mg==.ayEhMz0y0vNWB8_La4uuWt8elQhVcDjmOFbZvdPkpjIYEvhMH1BHEE1us1b4XLD4i5qdnp-CssyJ50SNg-WYHHxqF48QhmW2FyBTOCRRjs4t7ZD-CNyHdha85_1QAp3or-BgBTKUaUn4fFamMJW-6V4_huxWMHP8iK__QueLI_HeLcZHdfXkHsottM4iBjENT2gKKzCBZUKL59BZmPeldtBqN098E5pleXFI1LhOFEQldlqFakl-wrqw-8q-dFisBf5Ht8CAbWiQ7BQT5I6n4FQe7afwR_vuj-u4WLsCYZ_F71m9OxV83er2fQ2W4S4O2nj4F6_9CIOyXCWgf_6XSg

So, as the things stand, I am doing the following which has resulted in a verified signature. 

a. encrypting payload using symmetric key (not using Jose library)
b. using the encrypted payload to generate JWT with EncodePayload option as False.
c. send the detached JWT (currently manually taking out the payload from the token) and encrypted payload separately
d. On the receiving side - constructing the JWT back by concatenating the header , encrypted payload and signature 
e. Verify the JWT as constructed in (d) above. 
f. Decrypt the payload (not use Jose library) for further processing. 

This seems to work for me internally - I am in the process of carrying out the tests using a third party app / library. Hopefully now that I understand this - all should be good for implementation. 

Thanks for the handholding and being patient with me. 

--
